libpromises/policy_server.c

```diff
@@ -0,0 +1,182 @@
+ * @param new_policy_server can be 'host:port', same as policy_server.dat
+ */
+void PolicyServerSet(const char *new_policy_server)
+{
+    // Clean up static variables:
+    free(POLICY_SERVER);
+    free(POLICY_SERVER_HOST);
+    POLICY_SERVER         = NULL;
+    POLICY_SERVER_HOST    = NULL;
+
+    POLICY_SERVER_IP[0]   = '\0';
+    POLICY_SERVER_PORT[0] = '\0';
+    if (is_whitespace_empty(new_policy_server))
+    {
+        return;
+    }
+    else
+    {// Set POLICY_SERVER to be bootstrap argument/policy_server.dat contents
+        POLICY_SERVER = xstrdup(new_policy_server);
+    }
+
+    // Parse policy server in a separate buffer:
+    char *host_or_ip, *port;
+    char *buffer = xstrdup(new_policy_server);
+
+    AddressType address_type = ParseHostPort(buffer, &host_or_ip, &port);
+
+    if (address_type == ADDRESS_TYPE_OTHER)
+    {
+            POLICY_SERVER_HOST = xstrdup(host_or_ip);
+    }
+    else // ADDRESS_TYPE_IPV4 or ADDRESS_TYPE_IPV6
+    {
+        assert(strlen(host_or_ip) < CF_MAX_IP_LEN);
+        strcpy(POLICY_SERVER_IP, host_or_ip);
+    }
+
+    if ( ! NULL_OR_EMPTY(port) )
+    {
+        if(strlen(port) < CF_MAX_PORT_LEN)
+        {
+            strcpy(POLICY_SERVER_PORT, port);
+        }
+        else
+        {
+            Log(LOG_LEVEL_WARNING,
+                "Too long port number in PolicyServerSet: '%s'",
+                port);
+        }
+    }
+
+    free(buffer);
+}
+
+
+//*******************************************************************
+// POLICY SERVER GET FUNCTIONS:
+//*******************************************************************
+
+static char *CheckEmptyReturn(char *s)
+{
+    return (NULL_OR_EMPTY(s)) ? NULL : s;
+}
+
+/**
+ * @brief   Used to access the internal POLICY_SERVER variable.
+ * @return  Read-only string, can be 'host:port', same as policy_server.dat
+ */
+const char *PolicyServerGet()
+{
+    return POLICY_SERVER; // Don't use CheckedReturn, var should be NULL!
+}
+
+/**
+ * @brief   Gets the IP address of policy server, does lookup if necessary.
+ * @return  Read-only string, can be IPv4 or IPv6
+ */
+const char *PolicyServerGetIP()
+{
+    if (POLICY_SERVER_HOST == NULL)
+    {
+        return CheckEmptyReturn(POLICY_SERVER_IP);
+    }
+    assert(POLICY_SERVER_HOST[0] != '\0');
+    int ret = Hostname2IPString(POLICY_SERVER_IP, POLICY_SERVER_HOST,
+                                CF_MAX_IP_LEN);
+    if (ret != 0) // Lookup failed
+    {
+        return NULL;
+    }
+    return CheckEmptyReturn(POLICY_SERVER_IP);
+}
+
+/**
+ * @brief   Gets the host part of what was bootstrapped to (without port).
+ * @return  Read-only string, can be hostname or NULL (bootstrap to IP)
+ */
+const char *PolicyServerGetHost()
+{
+    return POLICY_SERVER_HOST; // Don't use CheckedReturn, var should be NULL!
+}
+
+/**
+ * @brief   Gets the port part of the policy server.
+ * @return  Read-only null terminated string of port number.
```

Comment:
Can  you mention in the doc comment that NULL is an acceptable value, if the port is not explicitly set? Same with the rest of the functions that this stands.